File: app/src/main/java/com/xeniac/warrantyroster/database/WarrantyRosterDatabase.java
package com.xeniac.warrantyroster.database;

import android.content.Context;

import androidx.room.Database;
import androidx.room.Room;
import androidx.room.RoomDatabase;

import com.xeniac.warrantyroster.Constants;

@Database(entities = {CategoryDataModel.class}, version = Constants.DB_VERSION)
public abstract class WarrantyRosterDatabase extends RoomDatabase {

    private static WarrantyRosterDatabase instance;

    public abstract CategoryDAO categoryDAO();

    public static WarrantyRosterDatabase getInstance(Context context) {
        if (instance == null) {
            instance = Room.databaseBuilder(context.getApplicationContext(),
                    WarrantyRosterDatabase.class, Constants.DB_FILE_NAME)
                    .allowMainThreadQueries()
                    .build();
        }
        return instance;
    }

    public static void destroyInstance() {
        instance = null;
    }
}

File: app/src/main/java/com/xeniac/warrantyroster_manager/db/WarrantyRosterDatabase.kt
package com.xeniac.warrantyroster_manager.db

import android.content.Context
import androidx.room.*
import com.xeniac.warrantyroster_manager.models.Category
import com.xeniac.warrantyroster_manager.util.Constants.Companion.DB_FILE_NAME_NEW

@Database(entities = [Category::class], exportSchema = true, version = 1)
@TypeConverters(MapConverter::class)
abstract class WarrantyRosterDatabase : RoomDatabase() {

    abstract fun getCategoryDao(): CategoryDao

    companion object {
        @Volatile
        private var instance: WarrantyRosterDatabase? = null
        private val LOCK = Any()

        operator fun invoke(context: Context) = instance ?: synchronized(LOCK) {
            instance ?: createDatabase(context).also { instance = it }
        }

        private fun createDatabase(context: Context) =
            Room.databaseBuilder(
                context.applicationContext,
                WarrantyRosterDatabase::class.java, DB_FILE_NAME_NEW
            ).build()
    }
}